### ISSUE TITLE
[OGE scheduler] Problem with the model reconstruction (carveme) step

### ISSUE THREAD

**Issue by shreyanshumale** [franciscozorrilla/metaGEM#50]
Hello, again Francisco!

So after the extractProteinbins step, I had 93 .faa files. I ran the 'bash metaGEM.sh -t carveme -j 93 -c 4 -m 20 -h 4' command after this and ended up with only 47 models in the GEMs folder (This is not due to the time limit as the job ended within an hour). I tried running the carveme step again but his time with 33 .faa files in the extractProteinbins folder. This time I ended up with only 19 models in the GEMs folder.

Please let me know if there is something that I am missing!

P.S. I tried running the carveme step separately on my PC and pasted the 93 files in the GEMs folder with the same naming scheme but then ran into errors when I tried running the memote or modelVis steps.

**Comment by shreyanshumale** [comment]
I think I have some clue about the cause of this problem. I am running metaGEM on a qsub managed cluster, and to my knowledge, there is no specific tag for the number of tasks (corresponding to the -j tag in metaGEM input). Please let me know your thoughts and if you know of a possible solution. I think the steps that are directly affected include carveme, memote, and smetana...

Thanks and regards,
Shreyansh

**Comment by franciscozorrilla** [comment]
Hey Shreyansh,

I suspect that the issues may be with the MAGs or the metagenomic dataset they originated from. Sometimes CarveMe doesn't get enough hits or "signal" from the input file to generate a metabolic model. This may be the case if your MAGs are low quality/highly fragmented/etc. Could you share the output log file for one of the failed CarveMe jobs? Please try the following:

1. Check the `nohup.out` file to find an example of a CarveMe job that failed (henceforth referred to as FAILED_MAG_ID)
2. Search the `logs/` folder for the file corresponding to the failed CarveMe job. Note that this folder may have hundreds of files, so you could search for the file as follows:
```
ls logs/|grep carve|grep FAILED_MAG_ID
```
3. Upload or paste the log file in this thread

Regarding the memote & modelVis errors, could you share more details about the error messages? If this is unrelated to the CarveMe problems then please feel free to open up a new issue.

Regarding cluster schedulers,`metaGEM` was built and optimized for SLURM clusters, as these are the only ones I have ever had access to. I am aware that this is an issue for some users (e.g. see issue #18), and I am happy to add support for non-SLURM clusters as required by users. Having said that, I am surprised that you were able to run/submit jobs at all. Did you make modifications to the `metaGEM.sh` script to be able to submit jobs on your SGE cluster?

Best wishes,
Francisco

**Comment by shreyanshumale** [comment]
Hello Francisco!

Firstly thanks for your quick replies!

Even I had the same doubts as you, which is why I ran carveme on the binned refined and reassembled MAGs on my PC as well. CarveMe ran smoothly, and I have models for all 93 of my MAGs. 
Upon checking the _nohup.out file_ I found that the following errors:
"_Unable to run job: job rejected: only three jobs are allowed per user (current job count: 3)
Exiting.
Error submitting jobscript (exit code 25):

Job failed, going on with independent jobs._"

That's why I think it's a _qsub_ problem. (Also note that all the commands before this do not require more than 2 jobs (I am following the unseen_bio tutorial


[modelVis.txt](https://github.com/franciscozorrilla/metaGEM/files/6592139/modelVis.txt)

[modelVis.txt](https://github.com/franciscozorrilla/metaGEM/files/6592165/modelVis.txt)
[nohup.txt](https://github.com/franciscozorrilla/metaGEM/files/6592166/nohup.txt)
[Snakefile.txt](https://github.com/franciscozorrilla/metaGEM/files/6592167/Snakefile.txt)
[SRR3577362.log.txt.log](https://github.com/franciscozorrilla/metaGEM/files/6592168/SRR3577362.log.txt.log)
[SRR3577362.memote.out.txt](https://github.com/franciscozorrilla/metaGEM/files/6592169/SRR3577362.memote.out.txt)


) to be submitted so didn't face any problems before this step)

I will attach the errors for memote and modelVis:

Yeah, I modified the scripts for my cluster. I am also attaching the modified _metaGEM.sh_ and _cluster_config.json_ files. (Note that I convert these to txt and attach because of GitHub restrictions).

Thanks and regards,
Shreyansh

**Comment by franciscozorrilla** [comment]
Hey Shreyansh,

No problem! I appreciate you posting your issues so that I can make `metaGEM` more accessible and user friendly :)

FYI, the early rules in the unseen bio tutorial are only using 2 jobs because there are only 2 samples in that dataset. Throughout most of the pipeline you will be submitting N jobs, where N = your number of samples.

I am curious though, how come you can only submit 3 jobs at a time? is this normal for SGE/OGE clusters? 

In any case, you may be able to avoid this issue by setting the qsub equivalent of `-j ` to be your MAX_JOB_NUM (in this case 3?). I would expect that this would result in only running 3 jobs at any given time and should eventually process all your MAGs into GEMs.

Thanks for sharing the modified `metaGEM.sh` script, I will have a look use this as a reference for adding support for SGE/OGE schedulers.

Best wishes,
Francisco

**Comment by shreyanshumale** [comment]
Hello Francisco!

Yeah, I realized that! In my case, I only had one sample, so thankfully did not come across any problems before the _carveme_ step. 
I am not sure if this is the case for all SGE/OGE clusters, but I did try to find the SGE equivalent of _-j_. Although, I haven't had any luck so far. I have written to the admin of our cluster if he has any insights into this problem; hopefully, I will have an answer soon!

For the carveme part if the multiple jobs approach doesn't workout (OGE/SGE) may I suggest using GNU parallel for this purpose?

Its just a simple command:

_ls ./protein_bins/*.faa |parallel --verbose "carve {} -o ./GEMs/{.}.xml --fbc2 --gapfill GM --mediadb ./media_db.tsv"_

>users can sue this on normal PCs also.
>to limit processor usage users can apply the _-j_ and _taskset_ arguments in the above line too.

I don't know if this is the correct/best approach but this has been working well for me so far!

Thanks,
Shreyansh
